pkg/storage/cacher.go

```diff
@@ -859,10 +859,19 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case !curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
 	}
+
+	// We don't want to block on c.result if c.done is already closed.
+	//
+	// TODO: This doesn't protect with the situation when c.result is
+	// already full and we are handing on it and in the meantime c.done
+	// is being closed and noone will clear the result channel.
+	// However, we can't heve select of two cases, as this may lead to
+	// delivering random part of events after c.done is closed but we
+	// are still sending them for a moment.
 	select {
-	case c.result <- watchEvent:
-	// don't block on c.result if c.done is closed
 	case <-c.done:
+	default:
```

Comment:
Sorry I wasn't clear enough. I agree this solves the problem that my PR is solving. However, this potentially also doesn't solve the problem that was supposed to be solved by the original PR which is leaking go-routines. By that I mean that if noone is draining the ResultChan, then we can still block on the ``` c.result <- watchEvent ``` forever in that case.

That said, I think that what you are suggesting is equally good as my PR and adds additional locking. Or am I missing something and what you suggested is somehow better?